File: ruby/lib/weft/facilitator/middleware.rb
require 'json'
require_relative 'client'

module Weft
  module Facilitator
    # Configuration for a route that requires payment.
    #
    #   {
    #     path: "/api/resource",          # string or Regexp
    #     price: "0.01",                  # price in asset units
    #     asset: "USDC",
    #     network: "base-sepolia",
    #     pay_to: "0x...",                # recipient address
    #     description: "Access resource",
    #     max_deadline_seconds: 60,
    #   }
    RouteConfig = Struct.new(
      :path, :price, :asset, :network, :pay_to,
      :description, :max_deadline_seconds,
      keyword_init: true
    ) do
      def initialize(**kwargs)
        super
        self.network ||= 'base-sepolia'
        self.max_deadline_seconds ||= 60
        self.description ||= ''
      end
    end

    class RackMiddleware
      PAYMENT_HEADERS = %w[HTTP_X_PAYMENT HTTP_PAYMENT_SIGNATURE].freeze
      X402_VERSION = 2

      # @param app [#call] The downstream Rack app
      # @param routes [Array<Hash>] Route payment configurations
      # @param config [Hash] Options including :facilitator_url
      def initialize(app, routes: [], config: {})
        @app = app
        @config = config
        @routes = routes.map do |r|
          r.is_a?(RouteConfig) ? r : RouteConfig.new(**r)
        end
        @client = Client.new(url: config[:facilitator_url])
      end

      def call(env)
        return @app.call(env) if @routes.empty?

        request_path = env['PATH_INFO'] || '/'
        route = match_route(request_path)

        return @app.call(env) unless route

        payment_header = extract_payment_header(env)

        return payment_required_response(route, env) unless payment_header

        # Verify the payment via facilitator
        payment_payload = parse_payment(payment_header)
        requirements = build_requirements(route, env)

        begin
          verify_result = @client.verify(
            payment_payload: payment_payload,
            payment_requirements: requirements
          )
        rescue StandardError => e
          return error_response(402, "Payment verification failed: #{e.message}")
        end

        unless verify_result['valid']
          return error_response(402, verify_result['message'] || 'Payment verification failed')
        end

        # Call the downstream app
        status, headers, body = @app.call(env)

        # If the app returned an error, don't settle
        return [status, headers, body] if status >= 400

        # Settle the payment
        begin
          settle_result = @client.settle(
            payment_payload: payment_payload,
            payment_requirements: requirements
          )

          return error_response(402, "Settlement failed: #{settle_result['message']}") unless settle_result['success']

          headers['X-Payment-TxHash'] = settle_result['txHash'] if settle_result['txHash']
        rescue StandardError => e
          return error_response(402, "Settlement failed: #{e.message}")
        end

        [status, headers, body]
      end

      private

      def match_route(path)
        @routes.find do |route|
          case route.path
          when Regexp
            route.path.match?(path)
          when '*'
            true
          else
            route.path == path
          end
        end
      end

      def extract_payment_header(env)
        PAYMENT_HEADERS.each do |header|
          value = env[header]
          return value if value && !value.empty?
        end
        nil
      end

      def parse_payment(header_value)
        JSON.parse(header_value)
      rescue JSON::ParserError
        # If not JSON, treat as an opaque token
        { 'token' => header_value }
      end

      def build_requirements(route, env)
        {
          'scheme' => 'exact',
          'network' => route.network,
          'asset' => route.asset,
          'amount' => route.price,
          'payTo' => route.pay_to,
          'maxTimeoutSeconds' => route.max_deadline_seconds,
          'resource' => {
            'url' => request_url(env),
            'method' => env['REQUEST_METHOD'],
            'description' => route.description
          }
        }
      end

      def request_url(env)
        scheme = env['rack.url_scheme'] || 'https'
        host = env['HTTP_HOST'] || env['SERVER_NAME'] || 'localhost'
        path = env['PATH_INFO'] || '/'
        "#{scheme}://#{host}#{path}"
      end

      def payment_required_response(route, env)
        body = {
          'x402Version' => X402_VERSION,
          'error' => 'Payment Required',
          'accepts' => [{
            'scheme' => 'exact',
            'network' => route.network,
            'asset' => route.asset,
            'amount' => route.price,
            'payTo' => route.pay_to,
            'maxTimeoutSeconds' => route.max_deadline_seconds,
            'resource' => {
              'url' => request_url(env),
              'method' => env['REQUEST_METHOD'],
              'description' => route.description
            }
          }]
        }

        [
          402,
          {
            'Content-Type' => 'application/json',
            'X-Payment-Required' => 'true'
          },
          [JSON.generate(body)]
        ]
      end

      def error_response(status, message)
        body = { 'error' => message }
        [
          status,
          { 'Content-Type' => 'application/json' },
          [JSON.generate(body)]
        ]
      end
    end
  end
end

File: go/facilitator/middleware.go
package facilitator

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

const x402Version = 2

// RoutePaymentConfig defines the payment requirements for a route.
type RoutePaymentConfig struct {
	// Price in asset units (e.g. "0.01")
	Price string
	// Asset symbol (e.g. "USDC")
	Asset string
	// Blockchain network (e.g. "base-sepolia")
	Network string
	// Recipient address
	PayTo string
	// Human-readable description of the resource
	Description string
	// Maximum deadline in seconds for the payment
	MaxDeadlineSeconds int
}

func (r *RoutePaymentConfig) setDefaults() {
	if r.Network == "" {
		r.Network = "base-sepolia"
	}
	if r.MaxDeadlineSeconds == 0 {
		r.MaxDeadlineSeconds = 60
	}
}

// RouteMatcher determines whether a request path requires payment.
type RouteMatcher func(path string) *RoutePaymentConfig

// ExactRoutes creates a RouteMatcher from a map of exact path → config.
func ExactRoutes(routes map[string]RoutePaymentConfig) RouteMatcher {
	return func(path string) *RoutePaymentConfig {
		if cfg, ok := routes[path]; ok {
			cfg.setDefaults()
			return &cfg
		}
		return nil
	}
}

// PrefixRoutes creates a RouteMatcher that matches path prefixes.
func PrefixRoutes(routes map[string]RoutePaymentConfig) RouteMatcher {
	return func(path string) *RoutePaymentConfig {
		for prefix, cfg := range routes {
			if strings.HasPrefix(path, prefix) {
				cfg.setDefaults()
				return &cfg
			}
		}
		return nil
	}
}

// MiddlewareConfig configures the x402 payment middleware.
type MiddlewareConfig struct {
	// Client is the facilitator client to use for verify/settle.
	Client *HTTPFacilitatorClient
	// Matcher determines which routes require payment.
	Matcher RouteMatcher
}

// paymentRequiredBody is the 402 response body.
type paymentRequiredBody struct {
	X402Version int                    `json:"x402Version"`
	Error       string                 `json:"error"`
	Accepts     []paymentRequirements  `json:"accepts"`
}

type paymentRequirements struct {
	Scheme            string        `json:"scheme"`
	Network           string        `json:"network"`
	Asset             string        `json:"asset"`
	Amount            string        `json:"amount"`
	PayTo             string        `json:"payTo"`
	MaxTimeoutSeconds int           `json:"maxTimeoutSeconds"`
	Resource          *resourceInfo `json:"resource,omitempty"`
}

type resourceInfo struct {
	URL         string `json:"url"`
	Method      string `json:"method"`
	Description string `json:"description,omitempty"`
}

// PaymentMiddleware returns an http.Handler that enforces x402 payments on
// configured routes. Requests to non-payment routes pass through directly.
//
// Usage:
//
//	client, _ := facilitator.NewFacilitatorClient(nil)
//	handler := facilitator.PaymentMiddleware(facilitator.MiddlewareConfig{
//	    Client: client,
//	    Matcher: facilitator.ExactRoutes(map[string]facilitator.RoutePaymentConfig{
//	        "/api/resource": {Price: "0.01", Asset: "USDC", Network: "base-sepolia", PayTo: "0x..."},
//	    }),
//	})(yourHandler)
func PaymentMiddleware(cfg MiddlewareConfig) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			route := cfg.Matcher(r.URL.Path)
			if route == nil {
				next.ServeHTTP(w, r)
				return
			}

			paymentHeader := r.Header.Get("X-Payment")
			if paymentHeader == "" {
				paymentHeader = r.Header.Get("Payment-Signature")
			}

			reqURL := requestURL(r)
			requirements := paymentRequirements{
				Scheme:            "exact",
				Network:           route.Network,
				Asset:             route.Asset,
				Amount:            route.Price,
				PayTo:             route.PayTo,
				MaxTimeoutSeconds: route.MaxDeadlineSeconds,
				Resource: &resourceInfo{
					URL:         reqURL,
					Method:      r.Method,
					Description: route.Description,
				},
			}

			if paymentHeader == "" {
				writePaymentRequired(w, requirements)
				return
			}

			// Parse payment payload from header
			var paymentPayload interface{}
			if err := json.Unmarshal([]byte(paymentHeader), &paymentPayload); err != nil {
				// Treat as opaque token
				paymentPayload = map[string]string{"token": paymentHeader}
			}

			// Verify payment
			ctx := r.Context()
			verifyResult, err := cfg.Client.Verify(ctx, paymentPayload, requirements)
			if err != nil {
				writeError(w, http.StatusPaymentRequired, fmt.Sprintf("Payment verification failed: %v", err))
				return
			}
			if !verifyResult.Valid {
				msg := verifyResult.Message
				if msg == "" {
					msg = "Payment verification failed"
				}
				writeError(w, http.StatusPaymentRequired, msg)
				return
			}

			// Buffer the response to check status before settling
			recorder := &responseRecorder{
				ResponseWriter: w,
				statusCode:     http.StatusOK,
			}

			next.ServeHTTP(recorder, r)

			// If downstream returned error, don't settle
			if recorder.statusCode >= 400 {
				return
			}

			// Settle the payment
			settleResult, err := cfg.Client.Settle(ctx, paymentPayload, requirements)
			if err != nil {
				// Response already sent; log but can't change status
				return
			}
			if settleResult.Success && settleResult.TxHash != "" {
				w.Header().Set("X-Payment-TxHash", settleResult.TxHash)
			}
		})
	}
}

// responseRecorder captures the status code from the downstream handler.
type responseRecorder struct {
	http.ResponseWriter
	statusCode int
	written    bool
}

func (rr *responseRecorder) WriteHeader(code int) {
	rr.statusCode = code
	rr.ResponseWriter.WriteHeader(code)
}

func (rr *responseRecorder) Write(b []byte) (int, error) {
	if !rr.written {
		rr.written = true
	}
	return rr.ResponseWriter.Write(b)
}

// Ensure responseRecorder implements http.Flusher if the underlying writer does.
func (rr *responseRecorder) Flush() {
	if f, ok := rr.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

func requestURL(r *http.Request) string {
	scheme := "https"
	if r.TLS == nil {
		scheme = "http"
	}
	host := r.Host
	if host == "" {
		host = "localhost"
	}
	return fmt.Sprintf("%s://%s%s", scheme, host, r.URL.Path)
}

func writePaymentRequired(w http.ResponseWriter, req paymentRequirements) {
	body := paymentRequiredBody{
		X402Version: x402Version,
		Error:       "Payment Required",
		Accepts:     []paymentRequirements{req},
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("X-Payment-Required", "true")
	w.WriteHeader(http.StatusPaymentRequired)
	json.NewEncoder(w).Encode(body)
}

func writeError(w http.ResponseWriter, status int, message string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(map[string]string{"error": message})
}

// Ensure the context has at minimum a background context.
func ensureContext(ctx context.Context) context.Context {
	if ctx == nil {
		return context.Background()
	}
	return ctx
}
